src/common/arrow/arrow_row_batch.cpp

```diff
@@ -106,82 +164,179 @@ void ArrowRowBatch::templateCopyNonNullValue<bool>(
 }
 
 template<>
-void ArrowRowBatch::templateCopyNonNullValue<std::string>(
-    ArrowVector* vector, Value* value, std::int64_t pos) {
+void ArrowRowBatch::templateCopyNonNullValue<STRING>(
+    ArrowVector* vector, const main::DataTypeInfo& typeInfo, Value* value, std::int64_t pos) {
     auto offsets = (std::uint32_t*)vector->data.data();
     auto strLength = value->strVal.length();
     offsets[pos + 1] = offsets[pos] + strLength;
     vector->overflow.resize(offsets[pos + 1]);
     std::memcpy(vector->overflow.data() + offsets[pos], value->strVal.data(), strLength);
 }
 
-void ArrowRowBatch::copyNonNullValue(ArrowVector* vector, Value* value, std::int64_t pos) {
-    switch (value->dataType.typeID) {
+template<>
+void ArrowRowBatch::templateCopyNonNullValue<LIST>(
+    ArrowVector* vector, const main::DataTypeInfo& typeInfo, Value* value, std::int64_t pos) {
+    vector->data.resize((pos + 2) * sizeof(std::uint32_t));
+    auto offsets = (std::uint32_t*)vector->data.data();
+    auto numElements = value->listVal.size();
+    offsets[pos + 1] = offsets[pos] + numElements;
+    auto numChildElements = offsets[pos + 1] + 1;
+    auto currentNumBytesForChildValidity = vector->childData[0]->validity.size();
+    auto numBytesForChildValidity = getNumBytesForBits(numChildElements);
+    vector->childData[0]->validity.resize(numBytesForChildValidity);
+    for (auto i = currentNumBytesForChildValidity; i < numBytesForChildValidity; i++) {
+        vector->childData[0]->validity.data()[i] = 0xFF;
+    }
+    if (typeInfo.childrenTypesInfo[0]->typeID != LIST) {
+        vector->childData[0]->data.resize(
+            numChildElements * Types::getDataTypeSize(typeInfo.childrenTypesInfo[0]->typeID));
+    }
+    for (auto i = 0u; i < numElements; i++) {
+        appendValue(
+            vector->childData[0].get(), *typeInfo.childrenTypesInfo[0], value->listVal[i].get());
+    }
+}
+
+template<>
+void ArrowRowBatch::templateCopyNonNullValue<NODE_ID>(
+    ArrowVector* vector, const main::DataTypeInfo& typeInfo, Value* value, std::int64_t pos) {
+    auto nodeID = value->getValue<nodeID_t>();
+    Value offsetVal((std::int64_t)nodeID.offset);
+    Value tableIDVal((std::int64_t)nodeID.tableID);
+    appendValue(vector->childData[0].get(), *typeInfo.childrenTypesInfo[0], &offsetVal);
+    appendValue(vector->childData[1].get(), *typeInfo.childrenTypesInfo[1], &tableIDVal);
+}
+
+template<>
+void ArrowRowBatch::templateCopyNonNullValue<NODE>(
+    ArrowVector* vector, const main::DataTypeInfo& typeInfo, Value* value, std::int64_t pos) {
+    appendValue(
```

Comment:
Why do we need to add the id and label manually? Could it just be a for-loop with 0 to start?

Reply:
Currently in NodeVal, properties are stored in separate fields from id and label, that's why we need to read them out separately as `value->nodeVal->getNodeIDVal()`, `value->nodeVal->getLabelVal()`, `value->nodeVal->getProperties()`